StateDB-based K8s fake client object tracker

joamaki opened the pull request:
This implements a new `testing.ObjectTracker` on top of StateDB that fixes the long-standing pain point with the k8s fake client, namely that its `Watch` implementation does not support multiple watchers and it will miss events that happen between `List` and `Watch`. This new implementation understands `ResourceVersion` and allows starting to watch from older versions without missing events.

This will also help doing more complicated integration tests where we would e.g. run the agent and the operator hives in the same process with a single fake client & tracker and we would be able to have both watching the same resources.

An extra bonus is that we now have debug logging for every k8s action that is done: 
```
logger.go:256: time=2025-05-08T18:35:43.964+02:00 level=DEBUG source=/home/jussi/go/src/github.com/cilium/cilium/pkg/k8s/client/object_tracker.go:324 msg=Get module=k8s-fake-client clientset=* gvr="cilium.io/v2, Resource=ciliumloadbalancerippools" k8sNamespace="" name=pool-a resourceVersion=9
logger.go:256: time=2025-05-08T18:35:43.964+02:00 level=DEBUG source=/home/jussi/go/src/github.com/cilium/cilium/pkg/k8s/client/object_tracker.go:436 msg=Update module=k8s-fake-client clientset=* obj="&{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:pool-a GenerateName: Namespace: SelfLink: UID: ResourceVersion:9 Generation:0 CreationTimestamp:0001-01-01 00:00:00 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ManagedFields:[]} Spec:{ServiceSelector:nil AllowFirstLastIPs: Blocks:[{Cidr:10.0.0.0/24 Start: Stop:}] Disabled:false} Status:{Conditions:[{Type:cilium.io/PoolConflict Status:False ObservedGeneration:0 LastTransitionTime:2025-05-08 18:35:43 +0200 CEST Reason:resolved Message:} {Type:cilium.io/IPsTotal Status:Unknown ObservedGeneration:0 LastTransitionTime:2025-05-08 18:35:43 +0200 CEST Reason:noreason Message:256} {Type:cilium.io/IPsAvailable Status:Unknown ObservedGeneration:0 LastTransitionTime:2025-05-08 18:35:43 +0200 CEST Reason:noreason Message:255}]}}" resourceVersion=9
etc.
```

joamaki committed:
k8s/client: StateDB-based ObjectTracker

The default testing.ObjectTracker that comes with client-go does
not properly implement the Watch() method:

  1) Does not support watching from a specific ResourceVersion
  2) Does not support multiple concurrent Watch() calls
  3) Does not support FieldSelector

This has been an ongoing pain as all the testing code needs to
make sure the implementation does not call Watch() too early.
This has required either populating the ObjectTracker before
starting the implementation or synchronizing with the implementation
to add objects only after Watch() has been called (e.g. db/initialized
or k8s/wait-watchers).

This commit implements a StateDB based ObjectTracker that properly
implements Watch() and FieldSelector.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer: Remove db/initialized, fix hostport.txtar

With the new fake client object tracker we no longer need to synchronize
with the reflectors before adding objects, so we can remove db/initialized.

Interestingly enough this change triggered the goleak to hit the workqueue
goroutine leak from Resource[T]. We don't have a way for waiting for that
background goroutine to exit so we'll just need to ignore it the same way
pkg/k8s/resource/resource_test.go does.

Fix the node name in hostport.txtar now that FieldSelector is implemented.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s/resource: Fix usage of fake client

Now that we're using our own object tracker the Tracker() method cannot be used
as that returns the wrong object tracker. It anyway makes more sense to go via the
typed API.

Fix the Resource[T] tests to not use Tracker() and also not to use ResourceVersion
as that's now correctly populated by the object tracker.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,617 @@
+	// domain to which this objects belongs, e.g. this is used
+	// to differentiate between the slim and kubernetes clientsets
+	// and avoid mixing them up since they have the same [gvr]
+	domain string
```
This is a bit clunky. The fake client uses "*" for everything except "kubernetes" for the core kubernetes clientsets. The reason for this is the overlap between slim and kubernetes and without this we'd mix them up. Could also have completely separate StateDB tables for each "clientset", but I kind of like just having one that is easy to inspect.

joamaki commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,617 @@
+	}, nil
+}
+
+type object struct {
```
FIXME add TableHeader and TableRow

joamaki commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,617 @@
+package client
+
+// Some parts of this code is originally from k8s.io/client-go/testing
+// Copyright 2015 The Kubernetes Authors.
```
Bunch of this code is copied from the original `testing.tracker`. Not sure how this should be attributed.

bimmlerd replied:
Maybe the community team has an answer?

joamaki replied:
Might perhaps be enough that the original code is already vendored and thus the license included. I'll still keep the disclaimer that code from here is based on the original.

bimmlerd commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,617 @@
+		return errs[0]
+	}
+	for _, obj := range list {
+		if err := s.Add(obj); err != nil {
```
I might be missing something, but when adding a list type, the call hierarchy is `Add -> addList -> Add`, but the both `Add`s want to create a WriteTxn on the table - doesn't this deadlock?

joamaki replied:
Sure does!

tommyp1ckles replied:
Guess I'm missing something here too 😆 
Is committing the elements of the list the correct behavior, instead of opening a write Tx for the duration ?

bimmlerd commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,617 @@
+}
+
+// Apply implements testing.ObjectTracker.
+func (s *statedbObjectTracker) Apply(gvr schema.GroupVersionResource, applyConfiguration runtime.Object, ns string, opts ...metav1.PatchOptions) error {
```
nit: for the k8s illiterate like me and given the ObjectTracker interface is somewhere else - could you add doc comments on what the semantics of these methods should be?

joamaki replied:
Sure! I'll also add some links to relevant code that explains how this fits in.

bimmlerd commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,617 @@
+	value := reflect.ValueOf(obj)
+	value.Elem().Set(reflect.New(value.Type().Elem()).Elem())
+
+	// For backward compatibility with behavior 1.30 and earlier, continue to handle apply
```
nit: I guess this comes from upstream - but do we need to retain compatibility with 1.30 and earlier? :sweat_smile:

joamaki replied:
Yeah that's upstream copypasta. I'll see if I can simplify.

bimmlerd commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,617 @@
+		logfields.Object, obj,
+		logfieldResourceVersion, version)
+
+	_, found, _ := s.tbl.Insert(wtxn, object{objectId: newObjectId(s.domain, gvr, ns, newMeta.GetName()), o: obj})
```
Could this be `tbl.Modify?`

joamaki replied:
It could, but we don't really care about the contents of the old object so `Insert` is simpler. Also `Modify` doesn't currently allow "not inserting", it'll anyway insert the new object but allows you to copy stuff from the old. Maybe should have a variant which we can do "UpdateIfExists". OTOH, `Get` + `Insert` performs quite ok.

bimmlerd commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,617 @@
+		logfieldResourceVersion, version)
+
+	_, found, _ := s.tbl.Insert(wtxn, object{objectId: newObjectId(s.domain, gvr, ns, newMeta.GetName()), o: obj})
+	if !found {
```
I'm suprised that the semantics aren't upsert - I suppose this is from upstream?

joamaki replied:
Yeah that's upstream semantics

bimmlerd commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,617 @@
+	}
+	wtxn := s.db.WriteTxn(s.tbl)
+	version := s.tbl.Revision(wtxn) + 1
+	newMeta.SetResourceVersion(strconv.FormatUint(version, 10))
```
Do we somehow need to be careful not to increment the version if the object hasn't changed - as in, does Update even get called with unchanged objects?

joamaki replied:
Ah good question. No idea. Will look into it.

joamaki replied:
I think we don't need to care about this. For sure the client-go fake didn't care.

bimmlerd commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,617 @@
+		logfields.Name, name)
+
+	wtxn := s.db.WriteTxn(s.tbl)
+	_, found, _ := s.tbl.Delete(wtxn, object{objectId: newObjectId(s.domain, gvr, ns, name)})
```
Surprised that this isn't idempotent deletion - but I guess this comes from upstream?

joamaki replied:
Yep also upstream. I'm keeping all the same errors as there.

bimmlerd reviewed (commented):
nice! I've added a comment re a deadlock that I think should happen, but maybe I misunderstood something.

One more place to clean up might be the controlplane tests, where there's the whole infra for waiting for the watch to be established. Otherwise I can do that in a separate PR.

joamaki committed:
k8s: Remove k8s/wait-watchers command

With the proper object tracker we no longer need this.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s/client: Override the Tracker() in fake clientsets

Since we're using Tracker() method in bunch of tests make it point
to our ObjectTracker. Unfortunately there's no clean way to do this
so I need to just play tricks with unsafe.Pointer. At least the result
can be validated.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
ciliumenvoyconfig: Remove db/initialized from tests

No need to wait for watchers anymore so remove db/initialized calls
from tests.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on daemon/k8s/testdata/pod.txtar:
```diff
@@ -5,19 +5,9 @@
-# as the fake k8s client is dumb and would miss events between the List
-# and Watch calls.
 hive start
 db/initialized
```
shouldn't this `db/initialized` be removed too?

bimmlerd reviewed (approved):
LGTM modulo one missed `db/initialized`

tommyp1ckles commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,648 @@
+	return true
+}
+
+func getFieldPathValue(obj interface{}, fieldPath string) (string, error) {
```
nit: can be `interface{}` -> `any`

tommyp1ckles commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,648 @@
+		m.SetResourceVersion(strconv.FormatUint(rev, 10))
+		matchingObjects = append(matchingObjects, o)
+	}
+	m, _ := meta.ListAccessor(list)
```
Q: How do we know this is necessarily of kind list?

joamaki replied:
Because it was created as a list (see line 363). This code is pretty much the same as the upstream code.

tommyp1ckles commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,648 @@
+	var fieldSelector fields.Selector
+	version := uint64(0)
+	if len(opts) > 0 && opts[0].ResourceVersion != "" {
+		opt := opts[0]
```
I'd expect the behaviour of most implementations here to apply all possible opts?

joamaki replied:
There can only be zero or one options, see https://github.com/kubernetes/client-go/blob/18a1faa115ed571de5af3e8f0f9c02973769ceb3/testing/fixture.go#L357.

I guess I could pull in that assertion here as well?

tommyp1ckles commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,648 @@
+			return nil, err
+		}
+
+		if opt.FieldSelector != "" {
```
Shouldn't we create the change iterator following validation of field selectors, opt resource versions etc?

joamaki replied:
That is what we are doing? See call to `feed` below and what it does there.

tommyp1ckles commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,648 @@
+
+// Stop implements watch.Interface.
+func (w *statedbWatch) Stop() {
+	close(w.stop)
```
Nit: Looking through other implementations of this, usually Stop() is implemented to be idempotent to not crash if called twice (however the interface doesn't specify this so either way is probably fine).

joamaki replied:
Added a sync.Once wrapping.

tommyp1ckles reviewed (approved):
Looks good, just some small things here and there..

joamaki committed:
treewide: Remove db/initialized uses, fix field selector issues

As db/initialized is not necessary anymore just for waiting for reflector/informer
to sync, remove remaining uses of it.

Fix the remaining field selector issues.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

marseel commented on daemon/k8s/testdata/pod.txtar:
```diff
@@ -80,7 +69,7 @@ default/nginx
     "namespace": "default",
     "uid": "d96be1eb-51df-4383-87ab-bc29d5b42933",
-    "resourceVersion": "482057",
+    "resourceVersion": "1",
```
nit: I would remove
```
resourceVersion: "482057"
```
from `-- pod.yaml --`

marseel commented on pkg/k8s/client/object_tracker.go:
```diff
@@ -0,0 +1,652 @@
+		return err
+	}
+
+	version := s.tbl.Revision(wtxn) + 1
```
just fyi, normally Revision is shared across all resources

joamaki replied:
That is what's happening here, e.g. we have one table for all resources, so shared "revision counter". Or did you mean the other way around, e.g. each resource has its own revision counter?

marseel replied:
> That is what's happening here, e.g. we have one table for all resources, so shared "revision counter". Or did you mean the other way around, e.g. each resource has its own revision counter?

ah ok, just reading through the code briefly (not for the codeowners that I was requested) I thought it works differently. Thanks

marseel reviewed (approved):
nice, thanks!
